src/cloudant/index.py

```diff
@@ -0,0 +1,173 @@
+        self.options = options
+        self._ref = method_ref
+        self._page_size = options.pop("page_size", 100)
+        self._valid_args = ALL_ARGS
+
+    def _prepare_extras(self):
+        """
+        check that extra params are expected/valid
+        """
+        for k in self.options:
+            if k not in self._valid_args:
+                raise ValueError("Invalid argument: {0}".format(k))
+        extras = python_to_couch(self.options)
+        return extras
+
+    def __getitem__(self, key):
+        """
+        implement key access and slicing
+
+        """
+        extras = self._prepare_extras()
+        if isinstance(key, basestring):
+            data = self._ref(key=key, **extras)
+            return data['rows']
+
+        if isinstance(key, list):
+            data = self._ref(key=key, **extras)
+            return data['rows']
+
+        if isinstance(key, slice):
+            # slice is startkey and endkey if str or array
+            str_or_none_start = isinstance(
+                key.start, (basestring, list)
+                ) or key.start is None
+            str_or_none_stop = isinstance(
+                key.stop, (basestring, list)
+                ) or key.stop is None
+            if str_or_none_start and str_or_none_stop:
+                # startkey/endkey
+                if key.start is not None and key.stop is not None:
+                    data = self._ref(
+                        startkey=key.start,
+                        endkey=key.stop,
+                        **extras
+                    )
+                if key.start is not None and key.stop is None:
+                    data = self._ref(startkey=key.start, **extras)
+                if key.start is None and key.stop is not None:
+                    data = self._ref(endkey=key.stop, **extras)
+                if key.start is None and key.stop is None:
+                    data = self._ref(**extras)
+                return data['rows']
+            # slice is skip:limit if ints
+            int_or_none_start = isinstance(
+                key.start, (int)
+                ) or key.start is None
+            int_or_none_stop = isinstance(
+                key.stop, (int)
+                ) or key.stop is None
+            if int_or_none_start and int_or_none_stop:
+                if key.start is not None and key.stop is not None:
+                    data = self._ref(
+                        skip=key.start,
+                        limit=key.stop,
+                        **extras
+                    )
+                if key.start is not None and key.stop is None:
+                    data = self._ref(skip=key.start, **extras)
+                if key.start is None and key.stop is not None:
+                    data = self._ref(limit=key.stop, **extras)
+                # both None case handled above
+                return data['rows']
+        raise RuntimeError("wtf was {0}??".format(key))
```

Comment:
Yeah, related to above, I think I need to add a CloudantArgumentError or something like that for all this argument processing stuff. Would prefer to to that in another PR though as this one is already > 300 lines